src/Shared/Recruit.Vacancies.Client/Infrastructure/TableStore/TableStorageQueryStore.cs

```diff
@@ -0,0 +1,125 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Threading.Tasks;
+using Esfa.Recruit.Vacancies.Client.Infrastructure.QueryStore;
+using Esfa.Recruit.Vacancies.Client.Infrastructure.QueryStore.Projections;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Table;
+using Newtonsoft.Json;
+using Polly;
+
+namespace Esfa.Recruit.Vacancies.Client.Infrastructure.TableStore
+{
+    internal sealed class TableStorageQueryStore : TableStorageBase, IQueryStore
+    {
+        public CloudTableClient TableClient { get; }
+        public CloudTable CloudTable { get; }
+
+        public TableStorageQueryStore(ILogger<TableStorageQueryStore> logger, IOptions<TableStorageConnectionsDetails> details)
+            : base(logger, details)
+        {
+            var storageAccount = CloudStorageAccount.Parse(details.Value.ConnectionString);
+            TableClient = storageAccount.CreateCloudTableClient();
+            CloudTable = TableClient.GetTableReference(StorageTableNames.TableName);
+        }
+
+        public async Task<T> GetAsync<T>(string typeName, string key) where T : QueryProjectionBase
+        {
+            var retrieveOperation = TableOperation.Retrieve<QueryEntity>(typeName, key);
+            var result = await RetryPolicy.ExecuteAsync(context => CloudTable.ExecuteAsync(retrieveOperation), new Context(nameof(IQueryStore.GetAsync)));
+            var queryEntity = (QueryEntity)result.Result;
+            if (queryEntity != null)
+            {
+                var actualItem = JsonConvert.DeserializeObject<T>(queryEntity.JsonData);
+                return actualItem;
+            }
+            return null;
+        }
+
+        public Task UpsertAsync<T>(T item) where T : QueryProjectionBase
+        {
+            CheckIfTableExists();
```

Comment:
Do we need to do this check on every operation? Can we create this table as part of the deployment process like we do with the Cosmos Mongo collections and just presume the table does exist?